content/blog/2020-12-16-compcert/index.md

```diff
@@ -0,0 +1,249 @@
+
+### Limitations
+
+CompCert doesn't guarantee that the executables it generates simulate the input C99 program; C99 and the executable itself lack formal semantics. In reality, CompCert verifies that its compilation from its semantics for C99, *CSem*, to its semantics for assembly languages, *Asm*, preserves those semantics. 
+
+#### Nondeterminism in Languages
+
+Nondeterminism in the *source* language weakens backward simulations, as their proofs only need to work for one of the source transitions possible. C99 is nondeterministic; CompCert addresses this by employing a reduction strategy [CStrategy][cstrat] that chooses one of the source program's potential behaviors.
+
+Nondeterminism in the *target* language weakens forwards simulations, as there may be runtime target program behaviors that are not accounted for in the proof. CompCert avoids this problem by proving that its target languages are deterministic. 
+
+#### Diverging Source Programs
+
+If we wanted diverging source programs to produce diverging target programs, the compiler would probably be able to do fewer optimizations, such as the dead code elimination example in the twelfth pass. Alternatively, compilation could fail on more programs; this is also not great, since we'd like to support as many programs as we can. CompCert approaches this by embedding source program safety and defined behavior into the assumptions of its correctness theorem.
+
+### Correctness, Composed
+
+Each pass between Csem and the abstract assembly language is proved with a forward simulation, some with a measure, some with direct transition correspondences. 
+
+By composing these simulations, we get forward simulations from RTL to Asm, from Cminor to Asm, and from Cstrategy to Asm. 
+
+We can then derive backwards simulations for these three language pairs.
+
+The theorem of correctness then asserts that Csem is in backwards simulation with Cstrategy, and therefore in backwards simulation with Asm. 
+
+### What's Missing? 
+
+Short answer: a lot! CompCert doesn't say anything about the semantics of the preprocessed C code. Nothing about safety of the assembling or linking processes, either. It doesn't guarantee that the executable has no segfaults or null pointer exceptions. It doesn't assume compilation halts, since the OCaml and other unverified portions may not terminate. Neither the compiler nor the code it generates has promises for time or space efficiency. It can't reason about threaded programs. Whether due to CompCert's implementation or C's limitations, it's definitely fair to criticize CompCert for its lack of range. 
```

Comment:
Well, who knows what's fair, but I think it's probably at least as fair to say "stop yelling at CompCert; it has verified lots and lots of stuff! Can't do everything at once." In other words, there are plenty of unverified things and limited capabilities. But their claim of "practical" does actually seem to be true in that people actually use CompCert for stuff.

Reply:
I will frame this differently; let me know if you think it's better this time around!